ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java

```diff
@@ -233,101 +187,61 @@ public int getShiftedPos(int pos) {
-        private final GsfSemanticLayer layer;
-        private final int endOffset;
-        private SequenceElement nextElement;
-        private int nextElementStartOffset = Integer.MAX_VALUE;
 
-        GsfHighlightSequence(GsfSemanticLayer layer, Document doc, 
-                int startOffset, int endOffset, 
-                SortedSet<SequenceElement> colorings) {
-            this.layer = layer;
-            this.endOffset = endOffset;
-
-            SequenceElement.ComparisonItem fromInclusive = new SequenceElement.ComparisonItem(startOffset);
-            SortedSet<SequenceElement> subMap = colorings.tailSet(fromInclusive);
-            iterator = subMap.iterator();
-        }
-
-        private SequenceElement fetchElementFromIterator(boolean updateNextElementStartOffset) {
-            int seStartOffset;
-            SequenceElement se;
-            if (iterator != null && iterator.hasNext()) {
-                se = iterator.next();
-                seStartOffset = se.range.getStart();
-                if (LOG.isLoggable(Level.FINE)) {
-                    LOG.log(Level.FINE, "Fetched highlight <{0},{1}>\n", // NOI18N
-                            new Object[]{seStartOffset, se.range.getEnd()});
-                }
-                if (seStartOffset >= endOffset) {
-                    se = null;
-                    seStartOffset = Integer.MAX_VALUE;
-                    iterator = null;
-                }
-            } else {
-                se = null;
-                seStartOffset = Integer.MAX_VALUE;
-                iterator = null;
-            }
-            if (updateNextElementStartOffset) {
-                nextElementStartOffset = seStartOffset;
-            }
-            return se;
+        GsfHighlightSequence(Iterator<SequenceElement> it) {
+            iterator = it;
         }
 
         @Override
         public boolean moveNext() {
-            if (nextElement != null) {
-                element = nextElement;
-                nextElement = fetchElementFromIterator(true);
-            } else {
-                if ((element = fetchElementFromIterator(false)) != null) {
-                    nextElement = fetchElementFromIterator(true);
-                }
-            }
-            return (element != null);
+            element = iterator.hasNext() ? iterator.next() : null;
+            return element != null;
         }
 
         @Override
         public int getStartOffset() {
-            return (element != null)
-                    ? layer.getShiftedPos(element.range.getStart())
-                    : Integer.MAX_VALUE;
+            return getShiftedPos(element.range().getStart());
         }
 
         @Override
         public int getEndOffset() {
-            return (element != null)
-                    ? Math.min(layer.getShiftedPos(element.range.getEnd()), nextElementStartOffset)
-                    : Integer.MAX_VALUE;
+            return getShiftedPos(element.range().getEnd());
         }
```

Comment:
wondering if you could move the fix into its own commit? Even if its just the one liner which removes `Math.min`.

It would get lost among the other improvements. We do also have a rule in the reviewer guide which states to try to avoid obfuscating fixes with code refactoring. (I know I did break it myself too a few times, but I believe this case here is easy to split still)

Reply:
Well, I probably close this PR and create two new ones. One with the one-line fix, the other is with the code cleanups.

While fixing the issues on  `HighlightImp`, it turned out that it's not used at all.
I guess it is just an old implementation copied over from `...java.semantic` package (that one is used), and later it turned out CSL would not use token based Highlighting.

Reply:
your choice. Commit messages is all what is left in the repo history once merged. If they describe what they do -> all good. But if you want to split this into two PRs -> also good of course.